Add message for MRP scheduler summary + better exception handling

mdietrichc2c opened the pull request:
This PR is to improve on the way the MRP scheduler handles exceptions:
- in case of SQL exceptions (mostly concurrency issues), the error is not logged.
- any other error is logged twice (in the log and in a summary message).

At the end of the execution, the message is posted on a mail group called "MRP Scheduler summary".

odony commented:
Hi, thanks for your PR,

My comments from PR #3841 still apply here, but there are more issues with this new PR:

 - As explained on #3841, there is no valid reason why the procurement scheduler should crash in the middle of an execution, except because of bugs. Badly configured product must not crash it either, otherwise it means there is a bug that must be fixed. We always prefer fixing those bugs rather than catching + hiding them.
 - Your PR is targeting 7.0 but introduces a new feature and several changes in the API. This is not possible in stable series according to the [stable policy](https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean). Such a change can only land in a master branch.
 - The procurement system has significantly changed in 8.0 and master, but the basics remains: there is no reason for it to crash at any point except due to a bug, so there should still be no reason to handle these cases as normal situations. A product that is badly configured should only result in a procurement exception, which will not stop the scheduler. And there is already a menu to review procurement exceptions.

In any case, what specific exceptions are you trying to log here? Bad product configurations should never trigger this...

mdietrichc2c commented:
Hi @odony ,
I can understand the preference for "fixing bugs", but this is not as easy:
- first, the exception you have correctly identified in [purchase.py](https://github.com/odoo/odoo/blob/6758b4cf/addons/purchase/purchase.py#L1119) should return an exception, since you can create procurements manually and must validate its contents when creating/modifying them; the real issue is that, in an automated process like the scheduler (that can create procurements), this will just make it fail.
- second, as any exception will stop the scheduler, this means that for customers, one incorrect product configuration will not allow for any mass procurement validation. That is unacceptable when the customer must validate around 2000~ procurements per day (as that was the case when the error was returned to us).
- third, exceptions are logged, but the customer has no means to see why the scheduler failed (or if it even succeeded) via the interface. That is why I added this new messaging queue.
- lastly, even if we're only catching OperationalErrors, the whole flow is not covered for that: you can see in the diff that I also added exception catching around the `button_restart` workflow in `_procure_confirm`, and in `create_automatic_op` as well.

If you want, I can alter this PR to correct only the fourth point (non-catched OperationalErrors) on odoo/master, and add the rest of the developments (logging, messaging + exception handling) to OCA/OCB. The first part of this PR is already on that branch, see https://github.com/OCA/OCB/pull/94.
I'll also see about the possibility to add this code in the master branch later on.

Best regards

StefanRijnhart commented on addons/procurement/schedulers.py:
```diff
@@ -39,9 +43,23 @@ def run_scheduler(self, cr, uid, automatic=False, use_new_cursor=False, context=
+        data_obj = self.pool['ir.model.data']
+        mail_group_obj = self.pool['mail.group']
+        __, mail_group_id = data_obj.get_object_reference(
+            cr, uid, 'procurement', 'group_mrp_scheduler_status')
```
Get_object_reference will raise at this point if the code is updated and Odoo is restarted but the module is not updated with the new mail group record. Can you at least catch this error and work around it?

As is confirmed on the upstream PR, this should be unacceptible to Odoo 7.0. Can you follow up Olivier's advice to try again with a PR on master?

StefanRijnhart commented on addons/procurement/procurement_data.xml:
```diff
@@ -28,5 +28,11 @@
+        <record model="mail.group" id="group_mrp_scheduler_status">
+            <field name="name">MRP scheduler summary</field>
+            <field name="group_ids" eval="[(4, ref('stock.group_stock_manager'))]"/>
+            <field name="description">Summary of the MRP scheduler</field>
```
For new terms like this, it is essential that we have the translation process set up, for instance with a specific OCB translation module. I asked @ccomb to do that in #197

Yenthe666 commented:
Hi @mdietrichc2c,

Thanks a lot for your PR!
Sadly we do not longer support version 7.0, since we only support up to three version back (8.0, 9.0 and 10.0).
If you can still reproduce this on a newer version (8.0, 9.0 or 10.0) then please create a new PR against one of these versions.
Sorry that we didn't handle this PR in time!

Regards,
Yenthe
